libqtile/backend/wayland/core.py

```diff
@@ -0,0 +1,346 @@
+        self.backend = Backend(self.display)
+        self.renderer = Renderer(self.backend, self.display)
+        self.compositor = Compositor(self.display, self.renderer)
+        self.socket = self.display.add_socket()
+        self.fd = None
+
+        # set up inputs
+        self.keyboards: List[keyboard.Keyboard] = []
+        self.grabbed_keys: List[Tuple[int, int]] = []
+        self.grabbed_buttons: List[Tuple[int, int]] = []
+        self.device_manager = DataDeviceManager(self.display)
+        self.seat = seat.Seat(self.display, "seat0")
+        self._on_request_set_selection_listener = Listener(self._on_request_set_selection)
+        self._on_new_input_listener = Listener(self._on_new_input)
+        self.seat.request_set_selection_event.add(self._on_request_set_selection_listener)
+        self.backend.new_input_event.add(self._on_new_input_listener)
+
+        # set up outputs
+        self.output_layout = OutputLayout()
+        self.outputs: List[output.Output] = []
+        self._on_new_output_listener = Listener(self._on_new_output)
+        self.backend.new_output_event.add(self._on_new_output_listener)
+
+        # set up cursor
+        self.cursor = Cursor(self.output_layout)
+        self.cursor_manager = XCursorManager(24)
+        self._on_request_cursor_listener = Listener(self._on_request_cursor)
+        self.seat.request_set_cursor_event.add(self._on_request_cursor_listener)
+        self._on_cursor_axis_listener = Listener(self._on_cursor_axis)
+        self._on_cursor_frame_listener = Listener(self._on_cursor_frame)
+        self._on_cursor_button_listener = Listener(self._on_cursor_button)
+        self._on_cursor_motion_listener = Listener(self._on_cursor_motion)
+        self._on_cursor_motion_absolute_listener = Listener(self._on_cursor_motion_absolute)
+        self.cursor.axis_event.add(self._on_cursor_axis_listener)
+        self.cursor.frame_event.add(self._on_cursor_frame_listener)
+        self.cursor.button_event.add(self._on_cursor_button_listener)
+        self.cursor.motion_event.add(self._on_cursor_motion_listener)
+        self.cursor.motion_absolute_event.add(self._on_cursor_motion_absolute_listener)
+
+        # set up shell
+        self.windows: List[window.Window] = []
+        self.xdg_shell = xdg_shell.XdgShell(self.display)
+        self._on_new_xdg_surface_listener = Listener(self._on_new_xdg_surface)
+        self.xdg_shell.new_surface_event.add(self._on_new_xdg_surface_listener)
+
+        # start
+        os.environ["WAYLAND_DISPLAY"] = self.socket.decode()
+        logger.info("Starting core with WAYLAND_DISPLAY=" + self.socket.decode())
+        self.backend.start()
+
+    def finalize(self):
+        self._on_new_xdg_surface_listener.remove()
+        self._on_request_cursor_listener.remove()
+        self._on_new_output_listener.remove()
+        self._on_new_input_listener.remove()
+        self._on_request_set_selection_listener.remove()
```

Comment:
Looks like the ones missing are ones I added second and forgot to remove here, so a helper to avoid that is probably a good idea. To be honest removing listeners when finalizing the core isn't as important as removing listeners when closing a window or removing an input or output. Either way, I might be able to put something together to make this more fool proof.

Reply:
I'm going to add a reminder to the To Do list to go over the listener handling code and see if there is a tidyer way to handle all of this, rather than add more to this PR.